### PR TITLE
Fix wrong reference to target instead of endpoint #1546

### LINKED ISSUE #1546: Modifying Task.uploadMultipart through endpointClosure has no effect
I'm trying to add authentication data (username/password) to an uploadMultipart task. First thing I tried was modifying `URLRequest.httpBody` in my authenticator plugin in `PluginType.prepare(_:target:)`, but this doesn't work for uploadMultipart tasks. Next, I tried modifying target.task in an endpointClosure like this:
```swift
        let endpointClosure: ((T) -> Endpoint<T>) = { target in
            var outtask = target.task
            if case .uploadMultipart(let multipart) = outtask {
              outtask = .uploadMultipart(multipart + [ /* Add authentication multipartData here */ ]
            }
            return Endpoint(url: target.path.absoluteString,
                            sampleResponseClosure: { .networkResponse(200, target.sampleData) },
                            method: target.method,
                            task: outtask,
                            httpHeaderFields: target.headers)
        }
```
This doesn't work either, since, if I understand MoyaProvider+Internal correctly, the original target gets passed to `performRequest(_:request:callbackQueue:progress:completion:endpoint:stubBehavior:)` alongside `endpoint`. Is there a way to modify multipart data outside of `TargetType` in Moya 10.0.1? Thanks in advance!

### PR DESCRIPTION
**Pull request by SD10**
This resolves #1546 by switching on `endpoint.task` instead of the original `target.task`. 

This way when a request is performed it will have any changes made to the: `.uploadMultipart`, `.uploadCompositeMultipart`, `.uploadFile`, `.downloadDestination`, and `.downloadParameters` tasks through an `endpointClosure`.

Thanks to @sunshinejr for setting up the test

### CODE REVIEW

File: Tests/EndpointClosureSpec.swift [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    var uploadMultipartFormData: Alamofire.MultipartFormData?
+
+    override func upload(multipartFormData: @escaping (Alamofire.MultipartFormData) -> Void, usingThreshold encodingMemoryThreshold: UInt64, with urlRequest: URLRequestConvertible, encodingCompletion: ((SessionManager.MultipartFormDataEncodingResult) -> Void)?) {
+        let uploadMultipartFormData = Alamofire.MultipartFormData()
+        multipartFormData(uploadMultipartFormData)
+        self.uploadMultipartFormData = uploadMultipartFormData
+    }
+
+    func convertMoyaMultipartFormDataToAlamofire(_ multipartBody: [Moya.MultipartFormData]) -> Alamofire.MultipartFormData {
```
Comment (sunshinejr) [inline Tests/EndpointClosureSpec.swift]: I've added this method here, but maybe we can add it as an internal extension to `Moya.MultipartFormData`?

Comment (SD10) [reply Tests/EndpointClosureSpec.swift]: I think it is better to keep this in the test target if we're not using it in the core library?

Comment (sunshinejr) [reply Tests/EndpointClosureSpec.swift]: [Actually, we use it in the core as well!](https://github.com/Moya/Moya/blob/f071d9703fbbd3f4a309fbbe32eae3ac6ca2f32d/Sources/Moya/MoyaProvider%2BInternal.swift#L154). If you have time to refactor this a little bit, go for it, if not, let's just make this an extension to `MultipartFormData` and add an issue for a little refactor?

Comment (SD10) [reply Tests/EndpointClosureSpec.swift]: Ok, thanks for pointing that out. I'll take care of it tonight 👍

**Comment by SD10** [comment]
@sunshinejr Sorry for the delay. I played around with the refactor in 91bb548. Let me know if that is what you were expecting. If not, we can kill the commit and make it an issue.

**Comment by sunshinejr** [comment]
@SD10 almost there! Everything is correct except we don't need `convertMoyaMultipartFormDataToAlamofire()` now 😉

**Comment by SD10** [comment]
@sunshinejr Fixed. Thanks for the eyes 👀

**Pull request merged**